tools/autograd/gen_python_functions.py

```diff
@@ -926,92 +1188,298 @@ def get_python_signature(declaration, include_out):
+            'type': 'bool',
+            'simple_type': 'bool',
+        }
+        python_binding_arguments.append(requires_grad_arg)
 
-    # add output arguments
-    name = declaration['name']
-    if name.endswith('_out'):
-        name = name[:-4]
-
-    if len(output_args) > 0 and include_out:
-        assert declaration['name'].endswith('_out')
-        if positional:
-            py_formal_args.append('*')
-            positional = False
-        typenames = [arg['simple_type'] for arg in output_args]
-        if len(typenames) > 1:
-            typename = 'TensorList[{}]'.format(len(typenames))
-        else:
-            typename = typenames[0]
-        if len(output_args) == 1:
-            # The nn module bindings are often not exposed to the user directly
-            # but via torch.nn modules and functionals.
-            py_formal_args.append(typename + ' ' + output_args[0]['name'] + '=None')
-        else:
-            # NB: For more than 1 output args the type name is a TensorList
-            # and as such we don't (yet) need to consider the naming.
-            py_formal_args.append(typename + ' out=None')
-
-    # we could put this in the loop above but we want to ensure both type dispatched args
-    # and python binding arguments are after the out argument; this matches the case
-    # where there is a python binding argument dtype, which is necessary to match
-    # the function signatures between the out and non-out variant.
-    assert len(type_args) <= 1
-    for arg in type_args:
-        if positional:  # assume type_args should be kwarg_only.
-            py_formal_args.append('*')
-            positional = False
-        py_formal_args.append(get_py_formal_arg(arg))
-
-    if len(declaration['python_binding_arguments']) > 0:
-        for arg in declaration['python_binding_arguments']:
-            if arg.get('kwarg_only', False) and positional:
-                py_formal_args.append('*')
-                positional = False
-            py_formal_args.append(get_py_formal_arg(arg))
+    return python_binding_arguments
 
-    # Python function signature.
-    # This is the string that we give to FunctionParameter, which is
-    # then parsed into the actual structure which we do parsing
-    # with.
-    return PYTHON_FUNCTION_SIGNATURE.substitute(name=name, py_formal_args=py_formal_args)
+#
+# declaration derived props, utils, etc.
+# declarations are dicts loaded from Declarations.yaml,
+# passed to our codegen methods by callers in gen_autograd
+#
+
+def is_tensor_self(arg):
+    return arg['name'] == 'self' and arg['simple_type'] == 'Tensor'
+
+
+def is_tensor_options(arg):
+    return arg['simple_type'] == 'TensorOptions'
+
+
+def is_scatter(arg):
+    return arg.get('scatter_args') is not None
+
+def is_output(arg):
+    return arg.get('output', False)
+
+
+def has_outputs(declaration):
+    return any([is_output(arg) for arg in declaration['arguments']])
+
+
+def get_tensor_options(declaration):
```

Comment:
any reason you didn't add type annotations to these?

Reply:
You mean like mypy-style? Didn't really consider it... I guess I probably wouldn't have thought to add them to some things but not the whole file. Or is that what you mean? Probably wouldn't want to stop and do everything but happy to do a subset if you want (I've seen them elsewhere in the binding CG but if there's a convention for when to do it, I missed it)

Reply:
Yes, I meant mypy-style.  I wasn't actually suggesting it, just curious if you thought about it.  There are some examples in function_wrapper.py and I find it makes it easier to read/reason about.